File: miscellaneous/palindrome_partitioning.cpp
//LeetCode - problem 1278

    int dp[101][101][101];
    int check(string s,int i,int j)
    {
        if(i>j) return 1000;
        int ans=0;
        while(i<j)
        {
            if(s[i]!=s[j])
                ans++;
            i++;j--;
        }
        return ans;
    }
    int solve(string s,int k,int i,int j)
    {
        if(i>j) return k<=0?0:1000;
        if(j-i+1==k) return dp[i][j][k]=0;
        if(j-i+1<k) return 1000;
        if(k==1) return dp[i][j][k]=check(s,i,j);
        if(dp[i][j][k]!=-1) return dp[i][j][k];
        int ans=1000;
        for(int x=i;x<j;x++)
        {
            int a=solve(s,1,i,x);
            int b=solve(s,k-1,x+1,j);
            ans=min(ans,a+b);
        }
        return dp[i][j][k]=ans;
    }
    int palindromePartition(string s, int k) {
        memset(dp,-1,sizeof(dp));
        return solve(s,k,0,s.size()-1);
    }

File: miscellaneous/palindrome_partitioning.py
'''
LeetCode - problem 1278
'''

	def palindromePartition(self, s: str, k: int) -> int:
		def calc(i,j):
			change=0
			while i<j:
				if s[i]!=s[j]:
					change+=1
				i+=1
				j-=1
			return change
		n=len(s)

		changes=[[0]*n for _ in range(n)]
		for i in range(n):
			for j in range(i+1,n):
				changes[i][j]=calc(i,j)

		dp=[[float('inf')]*n for _ in range(k)]
		for j in range(n):
			dp[0][j]=changes[0][j]
		for i in range(1,k):
			for r in range(i,n):
				for l in range(i,r+1):
					dp[i][r]=min(dp[i][r],dp[i-1][l-1]+changes[l][r])
		return dp[-1][-1]
